craft_parts/state_manager/states.py

```diff
@@ -54,6 +55,14 @@ def load_step_state(part: Part, step: Step) -> Optional[StepState]:
     with open(filename) as yaml_file:
         state_data = yaml.safe_load(yaml_file)
 
+    # fix project vars in loaded state data
+    opt = state_data.get("project-options")
+    if opt:
+        pvars = opt.get("project_vars")
+        if pvars:
+            for key, val in pvars.items():
+                pvars[key] = ProjectVar(**val)
```

Comment:
this might be working due to how the internals of dictionaries work, but I would prefer to see some for of reassignment here to state_data or a code comment.

Reply:
Expanded variable assignment and added extra comment. Ideally we should devise a way to add proper typing to the `project_options` definition so pydantic could automatically create the object.